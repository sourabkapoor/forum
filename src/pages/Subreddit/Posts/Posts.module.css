.container {
    border-radius: 4px;
    margin-bottom: 10px;
    overflow: hidden;
}

.container:hover {
    cursor: pointer;
}