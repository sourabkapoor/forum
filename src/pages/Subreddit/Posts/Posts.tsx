import {
  collection,
  doc,
  DocumentChange,
  getDoc,
  getDocs,
  limit,
  onSnapshot,
  orderBy,
  query,
  QuerySnapshot,
  where,
} from "firebase/firestore";
import { useEffect, useReducer, useState } from "react";
import Post from "../Post/Post";
import styles from "./Posts.module.css";
import { db, getUserId } from "../../../firebase";
import { DocumentSnapshot, DocumentData } from "firebase/firestore";
import {
  Link,
  Navigate,
  useLocation,
  useNavigate,
  useParams,
} from "react-router-dom";
import CSSModules from "react-css-modules";
import { useAppSelector } from "../../../hooks/hooks";
import AuthorsList from "../../../components/Skeletons/AuthorsList";

interface Props {
  posts?: DocumentData;
}

const Posts: React.FC<Props> = ({ posts }) => {
  const [randomPosts, setRandomPosts] = useState<DocumentData | undefined>(
    undefined
  );

  const location = useLocation();

  const { subredditName } = useParams();

  const navigate = useNavigate();

  useEffect(() => {
    if (location.pathname === "/user") return;
    if (!subredditName && posts !== undefined) return;
    else if (posts === undefined) {
      const postsRef = collection(db, "posts");

      const q = query(postsRef, orderBy("voteStatus", "desc"), limit(10));

      onSnapshot(q, (snapshot) => {
        const posts: DocumentData = [];

        snapshot.forEach((doc) => {
          posts.push(doc);
        });

        setRandomPosts(posts);
      });
    } else if (subredditName === undefined) {
      const postsRef = collection(db, "posts");

      const q = query(postsRef, where("subredditName", "==", subredditName));

      onSnapshot(q, (snapshot) => {
        const posts: DocumentData = [];

        snapshot.forEach((doc) => {
          posts.push(doc);
        });

        setRandomPosts(posts.slice(0, 10).sort(() => Math.random() - 0.5));
      });
    }
  }, [posts, subredditName, navigate, location.pathname]);

  if (posts === undefined && randomPosts === undefined) {
    return (
      <AuthorsList
        width="100%"
        animate={true}
        backgroundColor={"#333"}
        foregroundColor={"#999"}
        speed={1}
      />
    );
  }

  return (
    <>
      {(posts ?? randomPosts)?.map((doc: DocumentData) => {
        const data = doc.data();

        return (
          <Link
            key={doc.id}
            to={`/r/${data?.subredditName}/comments/${doc.id}`}
            state={{ ...data, id: doc.id }}
          >
            <Post key={doc.id} data={{ ...data, id: doc.id }} />
          </Link>
        );
      })}
    </>
  );
};

export default CSSModules(Posts, styles, {
  allowMultiple: true,
  handleNotFoundStyleName: "ignore",
});
