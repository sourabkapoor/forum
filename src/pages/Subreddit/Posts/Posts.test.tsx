import React from "react";
import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import userEvent from "@testing-library/user-event";
import each from "jest-each";
import Posts from "./Posts";

describe("Posts", () => {
  test("snapshot", () => {
    const { asFragment } = render(<Posts />);

    expect(asFragment()).toMatchSnapshot();
  });
});
