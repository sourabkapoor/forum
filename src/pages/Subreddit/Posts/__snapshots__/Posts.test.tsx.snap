// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Posts snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  />
</DocumentFragment>
`;
